src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -19,6 +21,153 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     internal partial class Binder
     {
+#nullable enable
+
+        internal enum EscapeLevel : uint
+        {
+            CallingMethod = Binder.CallingMethodScope,
+            ReturnOnly = Binder.ReturnOnlyScope,
+        }
+
+        /// <summary>
+        /// This is an argument that needs to be considered for MAMM rules. It's at least an 
+        /// output argument of the function and very often also an input. Essentially this 
+        /// represents all the writable ref and out arguments.
+        /// </summary>
+        internal readonly struct MixableDestination
+        {
+            internal BoundExpression Argument { get; }
+
+            /// <summary>
+            /// In the case this is the argument for a ref / out parameter this will refer
+            /// to the corresponding parameter. This will be null in cases like arguments 
+            /// passed to an arglist
+            /// </summary>
+            internal ParameterSymbol? Parameter { get; }
+
+            /// <summary>
+            /// This destination can only be written to by arguments that have at least 
+            /// this escape level. An destination that is <see cref="EscapeLevel.CallingMethod"/>
+            /// can never be written to by an argument that has a level of <see cref="EscapeLevel.ReturnOnly"/>
+            /// </summary>
+            internal EscapeLevel EscapeLevel { get; }
+
+            internal MixableDestination(ParameterSymbol parameter, BoundExpression argument)
+            {
+                Debug.Assert(parameter.RefKind.IsWritableReference());
+                Argument = argument;
+                Parameter = parameter;
+                EscapeLevel = parameter.RefKind == RefKind.Out ? EscapeLevel.ReturnOnly : EscapeLevel.CallingMethod;
```

Comment:
I'm somewhat reluctant to firmly tie those values together. Essentially silently converting the `uint` to `EscapeLevel`. I mostly used them as initializers for convenience. I gave a lot of thought to not doing taht.

Reply:
I think it would make sense to assert that the value lies in the expected range and convert it. Would be happy to chat more about this to make sure I understand.

Reply:
Just re-used `GetParameterValEscapeLevel` here.